filters/openpolicyagent/openpolicyagent.go

```diff
@@ -0,0 +1,318 @@
+	"github.com/zalando/skipper/filters/openpolicyagent/internal/util"
+)
+
+type OpenPolicyAgentFactory struct {
+	// Ideally share one Bundle storage across many OPA "instances".
+	// This allows to save memory on bundles that are shared
+	// between different policies (i.e. global team memberships)
+	// This not possible due to some limitations in OPA
+	// See https://github.com/open-policy-agent/opa/issues/5707
+	bundleStorage storage.Store
+}
+
+func NewOpenPolicyAgentFactory() OpenPolicyAgentFactory {
+	return OpenPolicyAgentFactory{
+		//BundleStorage: inmem.New(),
+	}
+}
+
+type OpenPolicyAgentInstanceConfig struct {
+	policyType             envoy.PolicyType
+	configTemplate         []byte
+	envoyContextExtensions map[string]string
+}
+
+func WithPolicyType(policyType envoy.PolicyType) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.policyType = policyType
+		return nil
+	}
+}
+
+func WithConfigTemplate(configTemplate []byte) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.configTemplate = configTemplate
+		return nil
+	}
+}
+
+func WithConfigTemplateFile(configTemplateFile string) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		var err error
+		cfg.configTemplate, err = os.ReadFile(configTemplateFile)
+		return err
+	}
+}
+
+func WithEnvoyContextExtensions(envoyContextExtensions map[string]string) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.envoyContextExtensions = envoyContextExtensions
+		return nil
+	}
+}
+
+func (cfg *OpenPolicyAgentInstanceConfig) GetPolicyType() envoy.PolicyType {
+	return cfg.policyType
+}
+
+func (cfg *OpenPolicyAgentInstanceConfig) GetEnvoyContextExtensions() map[string]string {
+	return cfg.envoyContextExtensions
+}
+
+func NewOpenPolicyAgentConfig(opts ...func(*OpenPolicyAgentInstanceConfig) error) (*OpenPolicyAgentInstanceConfig, error) {
+	cfg := OpenPolicyAgentInstanceConfig{}
+
+	cfg.policyType = envoy.IngressPolicyType
+	if val, found := os.LookupEnv("OPA_ENVOY_POLICY_TYPE"); found {
+		cfg.policyType = envoy.PolicyType(val)
+	}
+
+	for _, opt := range opts {
+		if err := opt(&cfg); err != nil {
+			return nil, err
+		}
+	}
+
+	err := cfg.policyType.IsValid()
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.configTemplate == nil {
+		cfg.configTemplate, err = os.ReadFile(configTemplate())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &cfg, nil
+}
+
+func configTemplate() string {
+	var configTemplate string
+	var found bool
+	if configTemplate, found = os.LookupEnv("OPA_CONFIG_TEMPLATE_FILE"); !found {
+		configTemplate = "opaconfig.yaml"
+	}
+
+	return configTemplate
+}
+
+func (factory *OpenPolicyAgentFactory) BundleStorage() storage.Store {
+	if factory.bundleStorage == nil {
+		return inmem.New()
```

Comment:
not sure how often this will be used. Does it makes sense to store it in `factory.bundleStorage`?

Reply:
Yes, that is where I want it as that would also share memory. There is an issue raised in OPA to support  this but it does not work today (See the comment on the factory). 

Factually this is called once per filter instance but agree that this is not clear from the code. Let me rework this a bit that the code makes sense today.

Reply:
Beware that in some clusters we have routing changes  every 10s. If we need to survive datastructures from one routing tree to the other we need to use a registry and hand-over the object instances by PostProcessors.
Examples are scheduler filters "fifo", "lifo" and shedder "admissionControl".
Right now I am not sure if we need this here. I am just reading the change for the first time.

Reply:
Interesting. Right now it would download the bundles again on a new route. It is completely stateless but 10s is only one order of magnitude more than the bootstrap.

Reply:
Sounds like we want to have a registry. If we would have 40k routes in a cluster and do this every 10s, it will be a bit huge.